# Assignment: Data Wrangling
### `! git clone https://github.com/ds4e/wrangling`
### Do Q1 and Q2, and either Q3 or Q4, for a total of 3 questions.

In [ ]:
%pip install pandas
%pip install numpy
%pip install seaborn

In [21]:
import pandas as pd
import numpy as np
import seaborn as sns

**Q1.** This question provides some practice cleaning variables which have common problems.

1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)

In [27]:
import os
os.getcwd()

'C:\\Users\\alexl\\AppData\\Local\\Programs\\Microsoft VS Code'

In [38]:
df = pd.read_csv('https://github.com/alaplace17/wrangling_updated/raw/main/assignment/data/airbnb_hw.csv', delimiter=',', on_bad_lines='skip')
print(df.shape, '\n')
df.head(5)

(30478, 13) 



Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0

In [40]:
price = df['Price']
price.unique()

array(['145', '37', '28', '199', '549', '149', '250', '90', '270', '290',
       '170', '59', '49', '68', '285', '75', '100', '150', '700', '125',
       '175', '40', '89', '95', '99', '499', '120', '79', '110', '180',
       '143', '230', '350', '135', '85', '60', '70', '55', '44', '200',
       '165', '115', '74', '84', '129', '50', '185', '80', '190', '140',
       '45', '65', '225', '600', '109', '1,990', '73', '240', '72', '105',
       '155', '160', '42', '132', '117', '295', '280', '159', '107', '69',
       '239', '220', '399', '130', '375', '585', '275', '139', '260',
       '35', '133', '300', '289', '179', '98', '195', '29', '27', '39',
       '249', '192', '142', '169', '1,000', '131', '138', '113', '122',
       '329', '101', '475', '238', '272', '308', '126', '235', '315',
       '248', '128', '56', '207', '450', '215', '210', '385', '445',
       '136', '247', '118', '77', '76', '92', '198', '205', '299', '222',
       '245', '104', '153', '349', '114', '320', '292', '22

In [45]:
price = df['Price']
price = price.str.replace(',', '') #replace commas with nothing
print(price.unique(), '\n')

['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1200' '158' '650' '234' '310' '695' '400' '166'
 '119' '62' '168' '340' '479' '43' '395' '144' '52' '47

In [46]:
price = pd.to_numeric(price, errors='coerce') # Change price datatype to float/numeric
print(price.unique(), '\n')

[  145    37    28   199   549   149   250    90   270   290   170    59
    49    68   285    75   100   150   700   125   175    40    89    95
    99   499   120    79   110   180   143   230   350   135    85    60
    70    55    44   200   165   115    74    84   129    50   185    80
   190   140    45    65   225   600   109  1990    73   240    72   105
   155   160    42   132   117   295   280   159   107    69   239   220
   399   130   375   585   275   139   260    35   133   300   289   179
    98   195    29    27    39   249   192   142   169  1000   131   138
   113   122   329   101   475   238   272   308   126   235   315   248
   128    56   207   450   215   210   385   445   136   247   118    77
    76    92   198   205   299   222   245   104   153   349   114   320
   292   226   420   500   325   307    78   265   108   123   189    32
    58    86   219   800   335    63   229   425    67    87  1200   158
   650   234   310   695   400   166   119    62   

In [48]:
print( 'Total missing: ', sum( price.isnull() ) ) # Checks to make sure no null values

Total missing:  0


2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 

In [67]:
url = 'https://github.com/alaplace17/wrangling_updated/blob/main/assignment/data/mn_police_use_of_force.csv?raw=true'
df = pd.read_csv(url)
df.head()

response_datetime               problem is_911_call primary_offense  \
0  2016/01/01 00:47:36  Assault in Progress          Yes          DASLT1   
1  2016/01/01 02:19:34                Fight           No          DISCON   
2  2016/01/01 02:19:34                Fight           No          DISCON   
3  2016/01/01 02:28:48                Fight           No          PRIORI   
4  2016/01/01 02:28:48                Fight           No          PRIORI   

  subject_injury         force_type   force_type_action   race     sex   age  \
0            NaN       Bodily Force  Body Weight to Pin  Black    Male  20.0   
1            NaN  Chemical Irritant       Personal Mace  Black  Female  27.0   
2            NaN  Chemical Irritant       Personal Mace  White  Female  23.0   
3            NaN  Chemical Irritant  Crowd Control Mace  Black    Male  20.0   
4            NaN  Chemical Irritant  Crowd Control Mace  Black    Male  20.0   

         type_resistance  precinct   neighborhood  
0                 Tensed         1  Downtown East  
1  Verbal Non-Compliance         1  Downtown West  
2  Verbal Non-Compliance         1  Downtown West  
3    Commission of Crime         1  Downtown West  
4    Commission of Crime         1  Downtown West

In [68]:
df.columns.tolist()

['response_datetime',
 'problem',
 'is_911_call',
 'primary_offense',
 'subject_injury',
 'force_type',
 'force_type_action',
 'race',
 'sex',
 'age',
 'type_resistance',
 'precinct',
 'neighborhood']

In [ ]:
df['subject_injury'].value_counts()


subject_injury
Yes    1631
No     1446
Name: count, dtype: int64

In [72]:
df['subject_injury'] = df['subject_injury'].fillna('Unknown/Missing')
print(df['subject_injury'].value_counts())

subject_injury
Unknown/Missing    9848
Yes                1631
No                 1446
Name: count, dtype: int64


There are 9848 missing values!

3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.

In [75]:
%pip install pyarrow

import pyarrow.parquet as pq

df = pd.read_parquet('https://github.com/alaplace17/wrangling_updated/blob/main/assignment/data/justice_data.parquet?raw=true')
df.head()

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


InternalStudyID REQ_REC# Defendant_Sex Defendant_Race Defendant_BirthYear  \
0        ADI00001        1             M              W                1986   
1        ADI00007        3             M              B                1956   
2        ADI00008        4             M              W                1990   
3        CDI00036        6             M              B                1989   
4        CDI00038        7             F              W                1988   

  Defendant_Age  Defendant_AgeGroup Defendant_AgeatCurrentArrest  \
0            31                   3                           31   
1            60                   6                           60   
2            27                   3                           27   
3            27                   3                           27   
4            28                   3                           28   

   Defendant_AttorneyTypeAtCaseClosure  Defendant_IndigencyStatus  ...  \
0                                   99                         99  ...   
1                                    9                          9  ...   
2                                    9                          9  ...   
3                                    0                          0  ...   
4                                    0                          0  ...   

  NewFelonySexualAssaultArrest_OffDate  \
0                                        
1                                        
2                                        
3                                        
4                                        

  NewFelonySexualAssaultArrest_ArrestDate  \
0                                           
1                                           
2                                           
3                                           
4                                           

   NewFelonySexualAssaultArrest_DaysBetweenContactEventandOffDate  \
0                                                                   
1                                                                   
2                                                                   
3                                                                   
4                                                                   

  NewFelonySexualAssaultArrest_DaysBetweenOffDateandArrestDate  \
0                                                999             
1                                                999             
2                                                999             
3                                                999             
4                                                999             

   NewFelonySexualAssaultArrest_DaysBetweenReleaseDateandOffDate  \
0                                                999               
1                                                999               
2                                                999               
3                                                999               
4                                                999               

   NewFelonySexualAssaultArrest_Disposition  \
0                                             
1                                             
2                                             
3                                             
4                                             

   Intertnalindicator_ReasonforExcludingFromFollowUpAnalysis  \
0                                                  4           
1                                                  5           
2                                                  5           
3                                                  5           
4                                                  0           

   CriminalHistoryRecordsReturnedorCMSRecordsFoundforIndividual  \
0                                                  1              
1                                                  1              
2                                                  1       

In [76]:
df.columns.tolist()

['InternalStudyID',
 'REQ_REC#',
 'Defendant_Sex',
 'Defendant_Race',
 'Defendant_BirthYear',
 'Defendant_Age',
 'Defendant_AgeGroup',
 'Defendant_AgeatCurrentArrest',
 'Defendant_AttorneyTypeAtCaseClosure',
 'Defendant_IndigencyStatus',
 'Defendant_RecordedZipCode_eMag',
 'Defendant_VirginiaResidencyStatus',
 'WhetherDefendantWasReleasedPretrial',
 'PretrialReleaseDate',
 'DaysBetweenContactEventAndPretrialRelease',
 'PretrialReleaseType1',
 'PretrialReleaseType2',
 'BondTypeAtInitialContact',
 'BondAmountAtInitialContact',
 'BondTypeAtRelease_v1',
 'BondTypeatRelease_v2',
 'BondAmountAtRelease',
 'WhetherDefendantReceivedPretrialServicesAgencySuperv_PTCC',
 'DaysBetweenReleaseandActivePretrialServicesAgencySupervDate',
 'DaysBetweenPretrialServicesAgencySupervReferralDateandSupervDate',
 'Indicator_PresumptiveDenialOfBail_19.2_120',
 'Indicator_ConditionsToBeReleasedSecuredBond_19.2_123',
 'IfReleasedonSecuredBond_TypeofSurety',
 'Indicator_BailTermSetByCourt_eMag',
 'AdditionalJailT

In [ ]:
release = df['WhetherDefendantWasReleasedPretrial']
print(release.unique(), '\n')
print(release.value_counts(), '\n')

[9 0 1] 

WhetherDefendantWasReleasedPretrial
1    19154
0     3801
9       31
Name: count, dtype: int64 



In [ ]:
release = release.replace(9,np.nan)
print(release.value_counts(), '\n')
sum(release.isnull()) #31 missing values

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64 



31

In [83]:
df['WhetherDefendantWasReleasedPretrial'] = release #replace column with new cleaned values
del release #delete release variable
df['WhetherDefendantWasReleasedPretrial'].value_counts() #check to make sure values are correct 

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64

4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [84]:
length = df['ImposedSentenceAllChargeInContactEvent']
type = df['SentenceTypeAllChargesAtConvictionInContactEvent']

print(length.unique() , '\n')

[' ' '60' '12' '.985626283367556' '36' '6' '24' '5.91375770020534' '120'
 '72' '11.9917864476386' '0' '2.95687885010267' '84' '108' '300' '240'
 '180' '4' '96' '2' '54' '.328542094455852' '44' '5' '115' '132' '48'
 '258' '34' '76' '.164271047227926' '.131416837782341' '111' '9' '3'
 '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '35' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '52' '30' '20'
 '192' '702' '14' '55' '53' '11.9055441478439' '114' '35.0061601642711'
 '68' '.657084188911704' '46.6242299794661' '102' '65' '200' '57'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '1500' '40' '284' '11' '118' '42' '162' '156'
 '47.2956878850103' '105' '51' '246' '29' '75' '324' '360'
 '34.4804928131417' '120.328542094456' '59.926078

In [85]:
length = pd.to_numeric(length, errors='coerce') # Change values to numeric
length_NA = length.isnull() # Creates missing dummy
print( np.sum(length_NA), '\n') # 9053 missing values

9053 



In [86]:
print(pd.crosstab(length_NA, type), '\n') # Category 4 is cases where charges were dismissed

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 



In [87]:
length = length.mask( type == 4,0) # Replace missing values with 0 for cases where charges were dismissed
length = length.mask( type == 9, np.nan) # Replace missing values with NaN for cases where sentence type is unknown

In [88]:
length_NA = length.isnull() # Creates new missing dummy
print( pd.crosstab(length_NA, type), '\n')
print( np.sum(length_NA), '\n') # 274 missing

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914  8779    0
True                                                 0     0    0     0  274 

274 



In [89]:
df['ImposedSentenceAllChargeInContactEvent'] = length # Replace column with cleaned values
del length, type # Destroy temporary vector for length and type
df['ImposedSentenceAllChargeInContactEvent'].value_counts()

ImposedSentenceAllChargeInContactEvent
0.000000      13732
12.000000      1404
0.985626       1051
6.000000        809
3.000000        787
              ...  
2.197125          1
21.164271         1
1.657084          1
2.299795          1
180.985626        1
Name: count, Length: 483, dtype: int64

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks.

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?
9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)

**Q3.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  ***This paper is about data tidying and how it is one of the most important aspects of data analysis. The specific structure of a tidy dataset is each variable is acolumn, each observation a row, and each type of observational unit is a table. Principles of tidy data provide standard ways to organise data values within a dataset.***
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  ***It is intended to make initial data cleaning easier because you don't have to start from scratch/reinvent the wheel each time. It is designed to facilitate initial exploration and analysis of the data and to simplify development of data analysis tools that worktogether.***
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  ***The first sentence means that tidy datasets follow a similar structure that allows for ease of analysis, while messy datasets are all uniquely messy and may have very different things wrong with them. The second sentence means that that it is intuitive to understand what a "dataframe/matrix" is made of but that it is harder to fully define what each variable and observation is and this can lead to uninitentional bad choices about cleaning and data organization.***
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  *** A dataset is a collection of values which are numeric or categorical/strings. Every value belogns to both a variable and an observaiton. A variable is a collection of values that measure the same attribute or property and an observation is a collection of values that measure it.***
  5. How is "Tidy Data" defined in section 2.3?
  ***Each variable is a column, each observation is a row, and each type of observational unit is a table. If data is not tidy, it is messy***
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  ***5 Most common problems with messy datasets: 1) Column headers are values, not variable names. 2) Multiple variables are stored in one column. 3) Variables are stored in both rows and columns. 4) Mulitple types of observational units are stored in the same table. 5) A single observational unit is stored in multiple tables. Table 4 is messy becaues the columns are really values of a "hidden" variable, which is income. Since income is actually a variable, you need a new column, income, alongside religion, and then frequency, as in table 6. Now, the columns are all the names of variables, rather than the values that variables take. Melting a dataset is this process of converting column-value variables into rows.***
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  ***Table 11 has days along the top, which are values. Table 12 melts those days into a single variable, date. That still isn't tidy, because the element variable contains variable names and not values --- tmax and tmin are measurements of the same day, which are really names of variables and not values themselves. Table 12(b) is tidy because all the entries are attributes and not variable names.***
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?
  ***If tidy data is only as useful as the tools that work with it, then tidy tools will be inextricably linked to tidy data. This makes it easy to get stuck in a local maxima where independently changing datat structures or data tools will not improve workflow. Wickham hopes that others will build on this framework to develop even better data stroage strategies and better tools.***

**Q4.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?